crates/worker/src/partition/snapshots/repository.rs

```diff
@@ -336,6 +359,171 @@ impl SnapshotRepository {
+            return Err(anyhow!(
+                "Unsupported snapshot format version: {:?}",
+                snapshot_metadata.version
+            ));
+        }
+
+        if snapshot_metadata.cluster_name != self.cluster_name {
+            panic!("Snapshot does not match the cluster name of latest snapshot at destination in snapshot id {}! Expected: cluster name=\"{}\", found: \"{}\"",
+                   snapshot_metadata.snapshot_id,
+                   self.cluster_name,
+                   snapshot_metadata.cluster_name);
+        }
+
+        // The snapshot ingest directory should be on the same filesystem as the partition store
+        // to minimize IO and disk space usage during import.
+        let snapshot_dir = TempDir::with_prefix_in(
+            format!("{}-", snapshot_metadata.snapshot_id),
+            &self.staging_dir,
+        )?;
+        debug!(
+            snapshot_id = %snapshot_metadata.snapshot_id,
+            path = ?snapshot_dir.path(),
+            "Getting snapshot data",
+        );
+
+        let directory = snapshot_dir.path().to_string_lossy().to_string();
+        let concurrency_limiter = Arc::new(Semaphore::new(DOWNLOAD_CONCURRENCY_LIMIT));
+        let mut downloads = JoinSet::new();
+        let mut task_handles = HashMap::with_capacity(snapshot_metadata.files.len());
+        for file in &mut snapshot_metadata.files {
+            let filename = file.name.trim_start_matches("/");
+            let expected_size = file.size;
+            let key = object_store::path::Path::from(format!(
+                "{prefix}{partition_id}/{path}/{filename}",
+                prefix = self.prefix,
+                partition_id = partition_id,
+                path = latest.path,
+                filename = filename,
+            ));
+            let file_path = snapshot_dir.path().join(filename);
+            let concurrency_limiter = Arc::clone(&concurrency_limiter);
+            let object_store = Arc::clone(&self.object_store);
+
+            let handle = downloads.build_task().name(filename).spawn(async move {
+                let _permit = concurrency_limiter.acquire().await?;
+                let mut file_data = StreamReader::new(
+                    object_store
+                        .get(&key)
+                        .await
+                        .map_err(|e| anyhow!("Failed to download snapshot file {:?}: {}", key, e))?
+                        .into_stream(),
+                );
+                let mut snapshot_file =
+                    tokio::fs::File::create_new(&file_path).await.map_err(|e| {
+                        anyhow!("Failed to create snapshot file {:?}: {}", file_path, e)
+                    })?;
+                let size = io::copy(&mut file_data, &mut snapshot_file)
+                    .await
+                    .map_err(|e| anyhow!("Failed to download snapshot file {:?}: {}", key, e))?;
+                if size != expected_size as u64 {
```

Comment:
nit: I don't this that this is problem here but a good practice is imo to use `u64::try_from(expected_size).expect("usize to fit into u64")`. That way we do observe overflows.